In [1]:
import pandas as pd
import numpy as np

# Data Preprocessing 

Load the grocery data set and show the first five samples.

In [39]:
data_grocery = pd.read_csv("../data/raw/breakfast_basket.csv")
data_grocery.head(5)

,City,Country,ISO_Country_Code,Region,Continent,Month,Month_Name,Item,Item_Key,Item_Category,...,Inflation_Source,FAO_Index_Value,FAO_Index_Date,FAO_YoY_Change_Pct,USDA_All_Food_Forecast_Pct,USDA_Food_At_Home_Pct,Data_Collection_Date,Source_URL,Population_Estimate,Breakfast_Basket_USD
0,New York,United States,USA,North America,North America,2025-10,October 2025,Milk (1 Liter),Milk_1L,Dairy,...,USDA Food Price Outlook / IMF WEO 2026,127.1,October 2025,-1.0,3.1,2.5,2026-03-20,https://www.numbeo.com/food-prices/in/New-York,8336817,16.18
1,New York,United States,USA,North America,North America,2025-11,November 2025,Milk (1 Liter),Milk_1L,Dairy,...,USDA Food Price Outlook / IMF WEO 2026,126.2,November 2025,-1.0,3.1,2.5,2026-03-20,https://www.numbeo.com/food-prices/in/New-York,8336817,16.40
2,New York,United States,USA,North America,North America,2025-12,December 2025,Milk (1 Liter),Milk_1L,Dairy,...,USDA Food Price Outlook / IMF WEO 2026,125.8,December 2025,-1.0,3.1,2.5,2026-03-20,https://www.numbeo.com/food-prices/in/New-York,8336817,17.02
3,New York,United States,USA,North America,North America,2026-01,January 2026,Milk (1 Liter),Milk_1L,Dairy,...,USDA Food Price Outlook / IMF WEO 2026,124.2,January 2026,-1.0,3.1,2.5,2026-03-20,https://www.numbeo.com/food-prices/in/New-York,8336817,16.80
4,New York,United States,USA,North America,North America,2026-02,February 2026,Milk (1 Liter),Milk_1L,Dairy,...,USDA Food Price Outlook / IMF WEO 2026,125.3,February 2026,-1.0,3.1,2.5,2026-03-20,https://www.numbeo.com/food-prices/in/New-York,8336817,16.82


Dimensions of the data frame /columns, rows/

In [34]:
data_grocery.shape

(10248, 27)

Get info for the data set.

In [40]:
data_grocery.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10248 entries, 0 to 10247
Data columns (total 27 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   City                        10248 non-null  object 
 1   Country                     10248 non-null  object 
 2   ISO_Country_Code            10248 non-null  object 
 3   Region                      10248 non-null  object 
 4   Continent                   10248 non-null  object 
 5   Month                       10248 non-null  object 
 6   Month_Name                  10248 non-null  object 
 7   Item                        10248 non-null  object 
 8   Item_Key                    10248 non-null  object 
 9   Item_Category               10248 non-null  object 
 10  Quantity                    10248 non-null  float64
 11  Unit                        10248 non-null  object 
 12  Price_Local                 10248 non-null  float64
 13  Currency_Local              102

There are no null values in columns.

The number of unique cities and continents.

In [41]:
len(data_grocery.City.unique()), len(data_grocery.Continent.unique())

(122, 6)

There are 122 cities in 6 continents.

List the names of the columns.

In [42]:
data_grocery.columns

Index(['City', 'Country', 'ISO_Country_Code', 'Region', 'Continent', 'Month',
       'Month_Name', 'Item', 'Item_Key', 'Item_Category', 'Quantity', 'Unit',
       'Price_Local', 'Currency_Local', 'Price_USD', 'Exchange_Rate',
       'YoY_Inflation_Estimate_Pct', 'Inflation_Source', 'FAO_Index_Value',
       'FAO_Index_Date', 'FAO_YoY_Change_Pct', 'USDA_All_Food_Forecast_Pct',
       'USDA_Food_At_Home_Pct', 'Data_Collection_Date', 'Source_URL',
       'Population_Estimate', 'Breakfast_Basket_USD'],
      dtype='object')

Check if "Breakfast_Basket_USD" is different for each sample for the same city at certain period.

In [51]:
data_grocery.loc[(data_grocery.Month == "2025-10") & (data_grocery.City == "New York")]["Breakfast_Basket_USD"].unique()

array([16.18])

We see that "Breakfast_Basket_USD" might be equal for the all samples in a certain month in any city. 

Since we need monthly average price, we could remove duplicates.

In [52]:
# first reamin only needed columns - "Month", "City", "Continent", Breakfast_Basket_USD"
data_grocery = data_grocery[["Month", "City", "Continent", "Breakfast_Basket_USD"]]
data_grocery

,Month,City,Continent,Breakfast_Basket_USD
0,2025-10,New York,North America,16.18
1,2025-11,New York,North America,16.40
2,2025-12,New York,North America,17.02
3,2026-01,New York,North America,16.80
4,2026-02,New York,North America,16.82
...,...,...,...,...
10243,2025-11,Perth,Oceania,12.65
10244,2025-12,Perth,Oceania,12.89
10245,2026-01,Perth,Oceania,12.55
10246,2026-02,Perth,Oceania,12.89


In [ ]:
# the number of duplicates
data_grocery.duplicated().sum()

np.int64(9516)

In [56]:
# remove duplicates
data_grocery.drop_duplicates()

,Month,City,Continent,Breakfast_Basket_USD
0,2025-10,New York,North America,16.18
1,2025-11,New York,North America,16.40
2,2025-12,New York,North America,17.02
3,2026-01,New York,North America,16.80
4,2026-02,New York,North America,16.82
...,...,...,...,...
10165,2025-11,Perth,Oceania,12.65
10166,2025-12,Perth,Oceania,12.89
10167,2026-01,Perth,Oceania,12.55
10168,2026-02,Perth,Oceania,12.89


For the study we need monthly average price of breakfast basket in each continent.

In [61]:
# group by Month and Continent, get the mean value of 'Breakfast_Basket_USD' column and reset index to have Month column
grocery_group_by_continent = data_grocery.groupby(by=["Month", "Continent"])[["Breakfast_Basket_USD"]].mean().reset_index()
grocery_group_by_continent

,Month,Continent,Breakfast_Basket_USD
0,2025-10,Africa,6.454545
1,2025-10,Asia,7.923256
2,2025-10,Europe,9.931951
3,2025-10,North America,14.518333
4,2025-10,Oceania,12.710000
5,2025-10,South America,7.974444
6,2025-11,Africa,6.571818
7,2025-11,Asia,8.025814
8,2025-11,Europe,10.047805
9,2025-11,North America,14.671667


For comparing the prices of oil and breakfast basket by month, we need a dataset that has index: "Month", columns: "Each continent" and "Oil" and values: the prices. 

In [63]:
# set "Month" column as index in the data set.
grocery_group_by_continent.set_index("Month")
grocery_group_by_continent

,Month,Continent,Breakfast_Basket_USD
0,2025-10,Africa,6.454545
1,2025-10,Asia,7.923256
2,2025-10,Europe,9.931951
3,2025-10,North America,14.518333
4,2025-10,Oceania,12.710000
5,2025-10,South America,7.974444
6,2025-11,Africa,6.571818
7,2025-11,Asia,8.025814
8,2025-11,Europe,10.047805
9,2025-11,North America,14.671667


In [64]:
# reshape the data, based on column values
grocery_group_by_continent = grocery_group_by_continent.pivot(index="Month", columns="Continent", values="Breakfast_Basket_USD")
grocery_group_by_continent

Continent,Africa,Asia,Europe,North America,Oceania,South America
Month,,,,,,
2025-10,6.454545,7.923256,9.931951,14.518333,12.710000,7.974444
2025-11,6.571818,8.025814,10.047805,14.671667,12.845000,8.101111
2025-12,6.652727,8.124884,10.191220,15.003333,13.116667,8.231111
2026-01,6.664545,8.066279,10.087073,14.825000,12.906667,8.154444
2026-02,6.818182,8.230698,10.271707,15.082500,13.155000,8.283333
2026-03,6.829091,8.228140,10.253415,15.054167,13.118333,8.300000


Store the data set for further usage.

In [65]:
data_grocery.to_csv("../data/processed/grocery_preprocessing.csv")